worldedit-bukkit/src/main/java/com/sk89q/worldedit/bukkit/adapter/Regenerator.java

```diff
@@ -0,0 +1,559 @@
+            tasks = new SequentialTasks(numlists);
+            for (int offset = 0; offset < numlists; offset++) {
+                ConcurrentTasks<SequentialTasks<Long>> para = new ConcurrentTasks((maxz - minz + 1) / numlists + 1);
+                for (int i = 0; minx + i * numlists + offset <= maxx; i++)
+                    para.add(byx.get(minx + i * numlists + offset));
+                tasks.add(para);
+            }
+        } else {
+            int numlists = Math.min(requiredneighbors * 2 + 1, maxz - minz + 1);
+
+            Int2ObjectOpenHashMap<SequentialTasks<Long>> byz = new Int2ObjectOpenHashMap();
+            int expectedListLength = (allcoords.size() + 1) / (maxz - minz);
+
+            //init lists
+            for (int i = minz; i <= maxz; i++) {
+                byz.put(i, new SequentialTasks(expectedListLength));
+            }
+
+            //sort into lists by x coord
+            for (Long xz : allcoords) {
+                byz.get(MathMan.unpairIntY(xz)).add(xz);
+            }
+
+            //create parallel tasks
+            tasks = new SequentialTasks(numlists);
+            for (int offset = 0; offset < numlists; offset++) {
+                ConcurrentTasks<SequentialTasks<Long>> para = new ConcurrentTasks((maxx - minx + 1) / numlists + 1);
+                for (int i = 0; minz + i * numlists + offset <= maxz; i++)
+                    para.add(byz.get(minz + i * numlists + offset));
+                tasks.add(para);
+            }
+        }
+
+        return tasks;
+    }
+
+    protected static Random getChunkRandom(long worldseed, int x, int z) {
+        Random random = new Random();
+        random.setSeed(worldseed);
+        long xRand = random.nextLong() / 2L * 2L + 1L;
+        long zRand = random.nextLong() / 2L * 2L + 1L;
+        random.setSeed((long) x * xRand + (long) z * zRand ^ worldseed);
+        return random;
+    }
+
+    //util
+    public static void time(Runnable r, String text) {
+        long starttask = System.currentTimeMillis();
+//            System.out.println(text);
+        r.run();
+        //Debug outputs:
+        //System.out.println(text + " took " + (System.currentTimeMillis() - starttask) + "ms");
+    }
+
+    //classes
+    public static abstract class ChunkStatusWrapper<IChunkAccess> {
+
+        /**
+         * Must not return a negative integer
+         *
+         * @return the radius of required neighbor chunk
+         */
+        public abstract int requiredNeigborChunkRadius();
+
+        public abstract String name();
+
+        public abstract void processChunk(Long xz, List<IChunkAccess> accessibleChunks);
+
+        void processChunkSave(Long xz, List<IChunkAccess> accessibleChunks) {
+            try {
+                processChunk(xz, accessibleChunks);
+            } catch (Exception e) {
+                System.err.println("Error while running " + name() + " on chunk " + MathMan.unpairIntX(xz) + "/" + MathMan.unpairIntY(xz));
```

Comment:
Should rather use a logger here

Reply:
which logger?

Reply:
`public static final Logger logger = LoggerFactory.getLogger(Regenerator.class);`

Reply:
done